[Ryan Ong Zheng Xun] iP

ryanozx opened the pull request:
# Atlas

## Introduction
> "I'm like a language model without a snooze button - always ready to chat and never in need of coffee!" - ChatGPT

Atlas helps you remember what to do so you don't have to. It's
1. text-based
2. easy and _fun_ to use
3. **REALLY** fast! :superhero:

It can
- store tasks, deadlines, and events
- list out your todo list
- mark your todo items as done
- save to your local disk
- load from your local disk

## Features
 - [x] Save to disk
 - [x] Load from disk
 - [ ] Change name
 - [ ] Implement GUI   

## Why the name?
[Atlas](https://en.wikipedia.org/wiki/Atlas_(mythology)) is a titan condemned to hold up the heavens or sky for eternity after the Titanonachy.

## Code dive
Atlas's code is straightforward. This is the `main` method that does it all:
```java
public static void main(String[] args) {
        new Duke("./data/", "duke.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

ryanozx committed:
Renamed bot and implemented greeting

ryanozx committed:
Added javadoc comments

ryanozx committed:
Added echo function

ryanozx committed:
Implemented text storing and listing

ryanozx committed:
Created Task class and implemented task status

ryanozx committed:
Added Todos, Deadlines, Events as task types

ryanozx committed:
Refactored and added more exceptions

ryanozx committed:
Implemented task deletion

ryanozx committed:
Created TaskTypes enums

ryanozx committed:
Add ability to save and load from file

ryanozx committed:
Merge branch 'branch-Level-7'

ryanozx committed:
Change date types and implement date command

ryanozx committed:
Merge branch 'branch-Level-8'

ryanozx committed:
Refactor code into multiple classes

ryanozx committed:
Merge branch 'A-MoreOOP'

ryanozx committed:
Divide classes into packages

ryanozx committed:
Merge branch 'branch-Packages'

ryanozx committed:
Merge branch 'branch-Packages' into add-gradle-support

ryanozx committed:
Set up gradle

ryanozx committed:
Merge branch 'add-gradle-support'

ryanozx committed:
Set up JUnit tests

ryanozx committed:
Merge branch 'branch-JUnit'

ryanozx committed:
Fix file creation bug

ryanozx committed:
Add JavaDoc comments

ryanozx committed:
Merge branch 'branch-A-JavaDoc'

ryanozx committed:
Enforce coding standards

ryanozx committed:
Merge branch 'branch-A-CodingStandard'

ryanozx committed:
Merge branch 'master' of https://github.com/ryanozx/ip into branch-A-CodingStandard

ryanozx committed:
Implement find command

ryanozx committed:
Merge branch 'branch-Level-9'

ryanozx committed:
Merge branch 'master' of https://github.com/ryanozx/ip into branch-Level-9

ryanozx committed:
Merge branch 'branch-A-CodingStandard'

ryanozx committed:
Merge branch 'branch-Level-9'

ryanozx committed:
Merge branch 'branch-Level-9' of https://github.com/ryanozx/ip into branch-Level-9

ryanozx committed:
Merge branch 'branch-Level-9'

ryanozx committed:
Add CheckStyle and fix style violations

ryanozx committed:
Merge branch 'branch-checkStyle'

ryanozx committed:
Create GUI for Atlas

ryanozx committed:
Merge branch 'branch-Level-10'

ryanozx committed:
Implement variable arguments for Mark and Unmark commands

ryanozx committed:
Merge branch 'branch-Varargs'

ncduy0303 commented on src/main/java/atlas/commands/Command.java:
```diff
@@ -0,0 +1,41 @@
+     * @param taskList TaskList for containing tasks
+     * @param storage Storage to handle file saving and loading
+     */
+    public abstract String execute(TaskList taskList, Storage storage);
```
Nice way of splitting the execute function for CLI and GUI.

ncduy0303 commented on src/main/java/atlas/commands/AddTaskCommand.java:
```diff
@@ -0,0 +1,50 @@
+        try {
+            taskList.addTask(task);
+            storage.save(taskList);
+            return String.format("Got it. I've added this task:\n"
```
Can consider storing these string messages as constants in the class.

ncduy0303 commented on src/main/java/atlas/commands/AddTaskCommand.java:
```diff
@@ -0,0 +1,50 @@
+        task = t;
+    }
+
+    @Override
```
Can consider adding {@InheritDoc} for these overidden methods

ncduy0303 commented on src/main/java/atlas/components/Parser.java:
I like how you split the parsing process into many methods of suitable length.

ncduy0303 commented on src/main/java/atlas/components/TaskList.java:
```diff
@@ -0,0 +1,110 @@
+     * @param idx One or more 0-based task indices
+     * @return Tasks marked as done
+     */
+    public List<Task> markTasks(int... idx) {
```
Can improve the variable name `idx` to signal that it is plural.

ncduy0303 reviewed (commented):
Overall, the code is clean and easy to follow. Great job!

owenyeo commented on src/main/java/atlas/Main.java:
```diff
@@ -0,0 +1,31 @@
+public class Main extends Application {
+    private final Atlas atlas = new Atlas("./data/", "atlas.txt");
+
+    @Override
```
Perhaps you could add docs for overriden methods using [@inheritDoc}?

owenyeo commented on src/main/java/atlas/commands/AddTaskCommand.java:
```diff
@@ -0,0 +1,50 @@
+/**
+ * Command to add task to task list
+ */
+public class AddTaskCommand extends Command {
```
I like how clear your class names are!

owenyeo commented on src/main/java/atlas/components/Parser.java:
```diff
@@ -0,0 +1,265 @@
+                return parseArguments(Command.Type.FIND, splitInput[1]);
+            default:
+                return new UnknownCommand();
+            }
```
I like how clean this is! Keep it up

owenyeo commented on src/main/java/atlas/tasks/Todo.java:
```diff
@@ -0,0 +1,27 @@
+ * Todo is a Task with no dates
+ */
+public class Todo extends Task {
+    /**
```
Perhaps an extra tab space can be added here?

owenyeo reviewed (commented):
Your code is clean looking and generally no naming issues as far as I can tell. Perhaps some improvements you could make is to ensure the block comment right after class declaration has a tab space before it, and you could be a little more elaborate in your JavaDoc. Overall, well done and keep it up!

ryanozx committed:
Add assert statements

There are no checks for whether the TaskList and Storage objects passed into
commands' execute() methods are non-null, causing potential undefined
behaviour and errors.

Adding assert statements to the execute() methods allow us to verify that
the TaskList and Storage objects passed in are non-null.

Assert statements will cause the program to terminate if encountered, and can be disabled to avoid affecting application performance.

ryanozx committed:
Refactor code to improve code quality

Command, Parser, and Storage violate SLAP principle, with multiple layers of abstractions within the same function.
Also, Storage has deeply nested functions.

Abstracting additional methods in these classes e.g. generateExecutionOutput() in Commands, parseTodoArgs(), parseDeadlineArgs(),
parseEventArgs() in Parser reduces the number of abstraction layers in one method.

Code quality is improved as methods no longer require knowledge of multiple layers of abstractions to perform an operation,
thereby reducing code fragility. This also results in shorter functions and improve readability.

Commands have duplicate behaviour (e.g. printing a list of tasks as the output).

Creating a new abstract class MultiTaskCommand that inherits from Command, and is in turn inherited by List commands, Mark and Unmark commands,
can reduce duplicate code. MultiTaskCommand::generateListOutput() prints the list of tasks.

Code quality is improved as common behaviour can be inherited instead of duplicated, reducing the number of lines of code and likelihood of
introducing bugs.

There are magic strings in the Commands (printed messages) and Parser (delimiters).

Assigning these strings to variables before using them can improve code readability.

Code quality is improved as other developers can better understand the purpose of the magic strings, enhancing readability and understanding.

ryanozx committed:
Merge pull request #1 from ryanozx/branch-A-Assertions

Add assert statements

ryanozx committed:
Merge branch 'branch-A-CodeQuality'

ryanozx committed:
Merge pull request #2 from ryanozx/branch-A-CodeQuality

Refactor code to improve code quality

ryanozx committed:
Add CI via GitHub Actions

ryanozx committed:
Merge branch 'branch-A-CI'

ryanozx committed:
Add style checks in CI

ryanozx committed:
Merge branch 'branch-A-CI'

ryanozx committed:
Merge branch 'master' of https://github.com/ryanozx/ip

ryanozx committed:
Implement ability to set and get reminders

There is no way to display a selection of tasks without specifying a specific date, hence users have to read through
the entire list of tasks to look for "important tasks".

Adding the ability to set and get reminders allows users to see just the important tasks (defined as tasks not complete
and are upcoming).

The option to set a reminder when creating a task hints to Atlas about the task's importance, thus users can be reminded
at the appropriate time.

ryanozx committed:
Merge pull request #3 from ryanozx/branch-BCD-Extension

Implement ability to set and get reminders

ryanozx committed:
Replace TaskList, Storage, and Parser method implementation with Streams

ryanozx committed:
Merge pull request #4 from ryanozx/branch-A-Streams

Replace TaskList, Storage, and Parser method implementation with Streams

ryanozx committed:
Improve GUI to scale with window resize and QoL improvements

Non-exhaustive list of GUI changes:
- Elements scale properly with window resizing
- Send button changed to circular button with triangle
- Send button colour changed to blue
- Send button colour changes upon hover for greater visual
  feedback
- Atlas's profile picture changed to one containing Atlas
- Font changed to Verdana for improved readability
- Font size increased to 18 for improved readability
- Newlines added in Atlas's lists for easier reading

ryanozx committed:
Merge pull request #6 from ryanozx/branch-BetterGui

Improve GUI to scale with window resize and QoL improvements

ryanozx committed:
Add personality to Atlas

Atlas's personality is inspired by its namesake in Greek mythology,
and draws some inspiration from the character in the God of War
franchise.

Atlas's response now contain more flavour, referring to tasks as
"labours" in a nod to Hercules' Twelve Labours.

A new help command has been created to display the list of possible
commands.

Error messages also contain flavour now, to reflect Atlas's
personality.

Atlas's response use the Courier New font, to distinguish from the
user's inputs.

ryanozx committed:
Fix checkstyle error

ryanozx committed:
Merge pull request #7 from ryanozx/branch-Personality

Branch personality

ryanozx committed:
Add more error handling with more specific exceptions

Previously, parser exceptions were handled at various levels in
the parser functions. Since we now need the exceptions to bubble
to the Atlas::getResponse() method, the exceptions are no longer
caught at earlier levels.

Custom exceptios have also been defined to provide greater
information in debugging, compared to the generic Java exceptions.

The parser now throws an exception if the reminder date is after
the deadline/end of event, while the Event constructor throws an
exception if the start of event is after the end of event.

ryanozx committed:
Merge pull request #8 from ryanozx/branch-MoreErrorHandling

Add more error handling with more specific exceptions

ryanozx committed:
Add screenshot of UI and title to window

ryanozx committed:
Merge pull request #9 from ryanozx/branch-ProjectWebsite

Add screenshot of UI and title to window

ryanozx committed:
Add user guide and replace Atlas's profile picture

ryanozx committed:
Merge pull request #10 from ryanozx/branch-UserGuide

Add user guide and replace Atlas's profile picture

ryanozx committed:
Fix broken tag
